bin/varnishd/cache/cache_conn_pool.c

```diff
@@ -249,17 +271,33 @@ VCP_Rel(struct conn_pool **cpp)
 		(void)shutdown(pfd->fd, SHUT_RDWR);
 		cp->n_kill++;
 	}
-	while (cp->n_kill) {
-		Lck_Unlock(&cp->mtx);
-		(void)usleep(20000);
-		Lck_Lock(&cp->mtx);
-	}
 	Lck_Unlock(&cp->mtx);
-	Lck_Delete(&cp->mtx);
-	AZ(cp->n_conn);
-	AZ(cp->n_kill);
-	free(cp->endpoint);
-	FREE_OBJ(cp);
+	if (cp->n_kill == 0) {
+		vcp_destroy(&cp);
+		return;
+	}
+	Lck_Lock(&dead_pools_mtx);
+	VRBT_INSERT(vrb, &dead_pools, cp);
+	Lck_Unlock(&dead_pools_mtx);
+}
+
+void
+VCP_RelPoll(void)
+{
+	struct conn_pool *cp, *cp2;
+
+	ASSERT_CLI();
+	Lck_Lock(&dead_pools_mtx);
+	VRBT_FOREACH_SAFE(cp, vrb, &dead_pools, cp2) {
```

Comment:
I still think we should take the dead_pools tree into a private tree here, try to destroy all pools on it outside the mtx and then put back any left overs.

Reply:
After discussing that with @dridi we think that this improvement could take place in a later PR, is that fine with you ?

Reply:
a `VRBT_HEAD()` is just a pointer to the root node, so swapping it is trivial. Inserting all remaining elements is a little less trivial, but no big deal either.

It is obvious how this PR implements a good improvement, but as we are already doing it, I do no see why we should have `VCP_Rel()` wait for `dead_pools_mtx`.

So I do not understand why we would need incremental commits rather than just getting this done.

Reply:
Ah, I see now that you added another unlock/lock around the destroy, so my argument about the lock is less relevant. Bus still this is unnecessary if we just swap the tree pointer.

Reply:
Alright I'll handle that

Reply:
Should be fine now